core/headerchain.go

```diff
@@ -349,8 +349,11 @@ func (hc *HeaderChain) GetAncestor(hash common.Hash, number, ancestor uint64, ma
 	}
 	for ancestor != 0 {
 		if rawdb.ReadCanonicalHash(hc.chainDb, number) == hash {
-			number -= ancestor
-			return rawdb.ReadCanonicalHash(hc.chainDb, number), number
+			ancestorHash := rawdb.ReadCanonicalHash(hc.chainDb, number-ancestor)
+			if rawdb.ReadCanonicalHash(hc.chainDb, number) == hash {
```

Comment:
I don't see what's different about line `353` and line `351`. Why is the same operation performed again?

Reply:
Because we are not locking the chain mutex. This ensures that the chain did not change between the two reads and `ancestorHash` is indeed an ancestor of `hash` because they are found in the same version of the canonical chain. I hope I assume correctly that if a reorg happens then the newer hash will change first.

Reply:
Ah, I guess you're doing this to make up for the data-race you introduced by dropping the mutex... 
Might be good to revamp this old pr, and make use of it in LES: https://github.com/ethereum/go-ethereum/pull/19319

Reply:
I like that PR, we should definitely use it as a speed-up layer on top of the hash chain in the db. LES needs to be able to access hashes and ancestors throughout the entire header chain though, and when we have to access the db we should not lock the mutex because the db access takes a long time and the mutex becomes a bottleneck (as we have seen in our tests). So I suggest to first do what I did in this PR and then replace the hash chain accessors with a cached function in a subsequent PR based on your work. Caching the canonical hashes is a big and critical change and should not be jammed in a bugfix PR.